_posts/2018-08-14-grpc-with-protobuf.md

```diff
@@ -0,0 +1,227 @@
+    ClientCall<CreateRequest, CreateResponse> call =
+        chan.newCall(KvGson.CREATE_METHOD, CallOptions.DEFAULT);
+    KvGson.CreateRequest req = new KvGson.CreateRequest();
+    req.key = key.array();
+    req.value = randomBytes(MEAN_VALUE_SIZE).array();
+
+    ListenableFuture<CreateResponse> res = ClientCalls.futureUnaryCall(call, req);
+    // ...
+```
+
+As you can see, we create a new `ClientCall` object from the `MethodDescriptor`, create the request, and then send it using `ClientCalls.futureUnaryCall` in the stub library.  gRPC takes care of the rest for us.  You can also make blocking stubs or async stubs instead of future stubs.
+
+
+## Receiving RPCs
+
+To update the server, we need to create a key-value service and implementation.  Recall that in gRPC, a _Server_ can handle one or more _Services_.  Again, what Protobuf would normally have generated for us we need to write ourselves.  Here is what the base service looks like:
+
+```java
+  static abstract class KeyValueServiceImplBase implements BindableService {
+    public abstract void create(
+        KvGson.CreateRequest request, StreamObserver<CreateResponse> responseObserver);
+
+    public abstract void retrieve(/*...*/);
+
+    public abstract void update(/*...*/);
+
+    public abstract void delete(/*...*/);
+
+    @Override
+    public final ServerServiceDefinition bindService() {
+      ServerServiceDefinition.Builder ssd = ServerServiceDefinition.builder(SERVICE_NAME);
+      ssd.addMethod(CREATE_METHOD, ServerCalls.asyncUnaryCall(
+          (request, responseObserver) -> create(request, responseObserver)));
+
+      ssd.addMethod(RETRIEVE_METHOD, /*...*/);
+      ssd.addMethod(UPDATE_METHOD, /*...*/);
+      ssd.addMethod(DELETE_METHOD, /*...*/);
+      return ssd.build();
+    }
+  }
+```
+
+`KeyValueServiceImplBase` will serve as both the service definition (which describes which methods the server can handle) and as the implementation (which describes what to do for each method).  It serves as the glue between gRPC and our application logic.  Practically no changes are needed to swap from Proto to GSON in the server code:
+
+```java
+final class KvService extends KvGson.KeyValueServiceImplBase {
+
+  @Override
+  public void create(
+      KvGson.CreateRequest request, StreamObserver<KvGson.CreateResponse> responseObserver) {
+    ByteBuffer key = ByteBuffer.wrap(request.key);
+    ByteBuffer value = ByteBuffer.wrap(request.value);
+    // ...
+  }
+```
+
+After implementing all the methods on the server, we now have a fully functioning gRPC Java, JSON encoding RPC system.  And to show you there is nothing up my sleeve:
+
+```sh
+$ ./gradlew :dependencies | grep -i proto
+$
+```
+
+## Optimizing the Code
+
+While GSON is not as fast as Protobuf, there's no sense in not picking the low hanging fruit.  Running the code the performance is pretty slow:
+
+```sh
+./gradlew installDist
+time ./build/install/kvstore/bin/kvstore
```

Comment:
I'm confused. Is there a benchmark in your code or something? Can you describe this benchmark?

Reply:
Yes, it was linked up at the top.  I added more detail there to indicate we are building on a previous example.